### PR TITLE
pull request :) 

### PR DESCRIPTION
**Pull request by smautner**
everything there now

### CODE REVIEW

**Comment by smautner** [comment]
ok i think i took care of all the things now.

File: graphlearn.py [hunk 1]
```diff
@@ -51,16 +54,23 @@
...
+class GraphLearnSampler:
     """
          wirte something
     """
 
-    def __init__(self):
-        self.bitmask = pow(2, 20) - 1
-        self.feasibility_checker = feasibility_checker()
-        self.vectorizer_expanded = myutils.my_vectorizer(complexity=3)
+    def __init__(self,radius_list=[],thickness_list=[],estimator=None,grammar=None,hash_bits=20):
```
Comment (fabriziocosta) [inline graphlearn.py]: use nbit for legacy with the rest of EDeN

File: graphlearn.py [hunk 2]
```diff
@@ -68,36 +78,64 @@ def save(self, filename):
...
-        self.grammar_functions.clean()
+        if not self.radius_list:
+            raise Exception ( "ERROR: tell me how to induce a grammar" )
 
 
-    def train_estimator(self, iterable_pos_train, n_jobs=-1, cv=10):
+        self.local_substitutable_graph_grammar = LocalSubstitutableGraphGrammar(self.radius_list, self.thickness_list,
+                                                                core_interface_pair_remove_threshold,
+                                                                interface_remove_threshold,
+                                                                hash_bitmask=self.hash_bitmask)
```
Comment (fabriziocosta) [inline graphlearn.py]: use nbits

File: graphlearn.py [hunk 3]
```diff
@@ -68,36 +78,64 @@ def save(self, filename):
...
-        l.sort()
-        element = int(len(l) * .1)
-        self.estimator.intercept_ -= l[element]
-        return self.estimator
+        estimator = graphlearn_utils.graphlearn_fit_estimator(positive_data_matrix=X_pos_train,
+                                                  negative_data_matrix=X_neg_train,
+                                                  cv=cv,
+                                                  n_jobs=n_jobs)
+
+        CUTOF=.1
```
Comment (fabriziocosta) [inline graphlearn.py]: its a parameter, called nu

File: graphlearn.py [hunk 4]
```diff
@@ -171,137 +217,185 @@ def sample_set(self, graph_iter, improvement_rules=
...
-
-        # ok lets go round and round
-        for x in xrange(self.improvement_rules['improvement_steps']):
+        sample_path=[graph]
+        accept_counter=0
+
+        # note to attach to the result in the end.
+        # currently it contains information on why we stopped the sampling before n_steps
+        special_notes='None'
+        for x in xrange(self.n_steps):
```
Comment (fabriziocosta) [inline graphlearn.py]: change x

File: graphlearn.py [hunk 5]
```diff
@@ -171,137 +217,185 @@ def sample_set(self, graph_iter, improvement_rules=
...
+        graph = graphlearn_utils.expand_edges(graph)
         self.score(graph)
+        self.similarity_checker(graph, set_reference=True)
+
+        if type(self.postprocessing)==str:
+            self.postprocessing = dill.loads(self.postprocessing)
+        return graph
+
+
+    def _sample(self, graph):
```
Comment (fabriziocosta) [inline graphlearn.py]: try to simplify _sample
it should just be:
while possible
  propose
  accept

File: graphlearn.py [hunk 6]
```diff
@@ -171,137 +217,185 @@ def sample_set(self, graph_iter, improvement_rules=
...
+            graph.score = self.estimator.predict_proba(transformed_graph)[0][1]
+            # print graph.score
+        return graph.score
+
+    def accept(self,graph_old,graph_new):
+        '''
+        returns the graph that was decided to be better, and an indicator if the better one is the new one
+        '''
 
+        graph_new = self.postprocessing(graph_new)
```
Comment (fabriziocosta) [inline graphlearn.py]: move out

File: graphlearn.py [hunk 7]
```diff
@@ -358,111 +468,239 @@ def choose_cip(self, graph):
 # ok moving this here instead of leaving it where it belongs prevents pickling errar ..
 #dont quite get it ...        
 def improve_loop_multi(x):
```
Comment (fabriziocosta) [inline graphlearn.py]: what?

File: graphlearn.py (hunk 1 above)
Comment (fabriziocosta) [inline graphlearn.py]: add defaults for radius thickness

File: graphlearn.py [hunk 8]
```diff
@@ -358,111 +468,239 @@ def choose_cip(self, graph):
...
 
 
 ################ALL THE THINGS HERE SERVE TO LEARN A GRAMMAR ############
 
 
 
-class core_interface_data:
-    def __init__(self, ihash=0, chash=0, graph=0, radius=0, thickness=0):
+class core_interface_pair:
+    def __init__(self, ihash=0, chash=0, graph=0, radius=0, thickness=0,core_nodes_count=0):
```
Comment (fabriziocosta) [inline graphlearn.py]: rename variables

**Pull request merged**